frontend/views/containers/chatroom/ChatMain.vue

```diff
@@ -891,9 +890,25 @@ export default ({
         if (!value) throw new Error('Unable to decrypt message')
 
         const isMessageAddedOrDeleted = (message: GIMessage) => {
-          if (![GIMessage.OP_ACTION_ENCRYPTED, GIMessage.OP_ACTION_UNENCRYPTED].includes(message.opType())) return {}
+          const allowedActionType = [GIMessage.OP_ACTION_ENCRYPTED, GIMessage.OP_ACTION_UNENCRYPTED]
+          let action = null
+          if (message.opType() === GIMessage.OP_ATOMIC) {
```

Comment:
Good catch with `OP_ATOMIC`. However, there could technically be multiple messages in OP_ATOMIC, so I think it'd be better to call the parent function recursively with all actions.

Reply:
For the future `OP_ATOMIC` messages,
- Could it have more than two child messages of `OP_ACTION_UNENCRYPTED` and `OP_ACTION_ENCRYPTED`?
- Could it create two messages? Then, there could be a serious problem to set message hash in `createMessage` function. Because can not create two messages with one hash.
- Could it delete messages and add messages at once?

Reply:
> Could it have more than two child messages of OP_ACTION_UNENCRYPTED and OP_ACTION_ENCRYPTED?

Yes, you _could_ have any number of actions. (We don't generate such messages in GI as far as I know, but they are _possible_)

> Could it create two messages? Then, there could be a serious problem to set message hash in createMessage function. Because can not create two messages with one hash.

I don't follow entirely, but I don't think it'd be an issue with `createMessage` because it doesn't create OP_ATOMIC with multiple messages. However, in general, many 'messages' can have the same hash if part of OP_ATOMIC. Maybe the hash could be `${hash}-${index}`?

> Could it delete messages and add messages at once?

From the protocol POV, OP_ATOMIC is just a wrapper with an array of opcodes, and it can contain any number of operations and opcodes, except OP_ATOMIC itself.